### PR TITLE
Adding Support for Desktop

### PR DESCRIPTION
**Pull request by rodydavis**
## Status
READY

## Breaking Changes
NO

## Description
Support for Desktop Directory to Story Application Data

## Related PRs
N/A

## Todos
- [x] Tests
- [x] Documentation
- [X] Examples

## Steps to Test or Reproduce
Outline the steps to test or reproduce the PR here.

```
Open project in VSCode and click run for MacOS, Linux or Windows
```

## Impact to Remaining Code Base
This PR will affect:

* Saving

### CODE REVIEW

**Review by felangel — CHANGES_REQUESTED** [review]
@AppleEducate thanks for the contribution! 🙏 Looks like some tests are failing let me know if you need any help 👍

**Comment by rodydavis** [comment]
@felangel  sure thing, ill look through the test today!

**Comment by rodydavis** [comment]
@felangel  the test run great on my machine. I am on the master channel and had to add `TestWidgetsFlutterBinding.ensureInitialized();` on the first line of the test main() as per the flutter documentation.

**Comment by rodydavis** [comment]
I think the CI test is not on the master channel

**Comment by rodydavis** [comment]
@felangel ok tests have passed

File: lib/src/hydrated_bloc_storage.dart [hunk 1]
```diff
@@ -67,4 +67,19 @@ class HydratedBlocStorage implements HydratedStorage {
     _instance = null;
     return await _file.exists() ? await _file.delete() : null;
   }
+
+  static Future<Directory> getDocumentDir({bool testing = false}) async {
+    if (!testing) {
```
Comment (felangel) [inline lib/src/hydrated_bloc_storage.dart]: why is this needed?

Comment (rodydavis) [reply lib/src/hydrated_bloc_storage.dart]: the testing method? Because it picks up the linux path of the circus CI linux directory and not the temp directory. On desktop when not testing it works great.

Comment (felangel) [reply lib/src/hydrated_bloc_storage.dart]: But with this approach it won't be possible to test the MacOS/Linux or Windows 😢 Can we instead fake the environments in the tests without changing the implementation?

Comment (rodydavis) [reply lib/src/hydrated_bloc_storage.dart]: Let me see what i can do!

Comment (rodydavis) [reply lib/src/hydrated_bloc_storage.dart]: @felangel this is the error im getting ` FileSystemException: Cannot open file, path = '/home/cirrus/.config/.hydrated_bloc.json' (OS Error: No such file or directory, errno = 2)` not sure what to do for testing.

Comment (felangel) [reply lib/src/hydrated_bloc_storage.dart]: I'll try to take a look later today and see if I have any luck 🤞

Comment (rodydavis) [reply lib/src/hydrated_bloc_storage.dart]: OK sweet!

Comment (felangel) [reply lib/src/hydrated_bloc_storage.dart]: I have a fix 😄
Can you add me as a collaborator to your fork?

**Comment by rodydavis** [comment]
Added!

File: lib/src/hydrated_bloc_storage.dart (hunk 1 above)
Comment (rodydavis) [reply lib/src/hydrated_bloc_storage.dart]: Done! ✅

Comment (felangel) [reply lib/src/hydrated_bloc_storage.dart]: Can you review https://github.com/AppleEducate/hydrated_bloc/pull/1?

Comment (rodydavis) [reply lib/src/hydrated_bloc_storage.dart]: Merged! 🚀

**Review by felangel — DISMISSED** [review]
💯

File: test/hydrated_bloc_delegate_test.dart [hunk 2]
```diff
@@ -27,11 +29,13 @@ void main() {
     storage = MockStorage();
     delegate = HydratedBlocDelegate(storage);
     bloc = MockBloc();
+    platform = FakePlatform(operatingSystem: 'ios');
```
Comment (felangel) [inline test/hydrated_bloc_delegate_test.dart]:
```suggestion
```

File: test/hydrated_bloc_delegate_test.dart [hunk 3]
```diff
@@ -14,6 +15,7 @@ void main() {
   MockStorage storage;
   HydratedBlocDelegate delegate;
   MockBloc bloc;
+  Platform platform;
```
Comment (felangel) [inline test/hydrated_bloc_delegate_test.dart]:
```suggestion
```

File: test/hydrated_bloc_delegate_test.dart [hunk 4]
```diff
@@ -1,10 +1,11 @@
 import 'dart:convert';
-import 'dart:io';
+import 'dart:io' hide Platform;
 import 'package:flutter/services.dart';
 import 'package:flutter_test/flutter_test.dart';
 import 'package:mockito/mockito.dart';
 import 'package:bloc/bloc.dart';
 import 'package:hydrated_bloc/hydrated_bloc.dart';
+import 'package:platform/platform.dart';
```
Comment (felangel) [inline test/hydrated_bloc_delegate_test.dart:8]:
```suggestion
```

File: test/hydrated_bloc_delegate_test.dart [hunk 5]
```diff
@@ -1,10 +1,11 @@
 import 'dart:convert';
-import 'dart:io';
+import 'dart:io' hide Platform;
```
Comment (felangel) [inline test/hydrated_bloc_delegate_test.dart]:
```suggestion
import 'dart:io';
```

**Review by felangel — DISMISSED** [review]
💯

**Review by felangel — DISMISSED** [review]

**Review by felangel — APPROVED** [review]
Thanks for the awesome contribution! 🙏

**Comment by rodydavis** [comment]
Awesome 😎 I’ll work on web support next 😉

File: lib/src/hydrated_bloc_storage.dart [hunk 6]
```diff
@@ -67,4 +70,17 @@ class HydratedBlocStorage implements HydratedStorage {
     _instance = null;
     return await _file.exists() ? await _file.delete() : null;
   }
+
+  static Future<Directory> _getDocumentDir(Platform platform) async {
+    if (platform.isMacOS || platform.isLinux) {
```
Comment (stuartmorgan-g) [inline lib/src/hydrated_bloc_storage.dart:75]: FYI, forcing this codepath will actually make your package behave worse for people using [a desktop path_provider implementation](https://github.com/google/flutter-desktop-embedding/tree/master/plugins/flutter_plugins/path_provider_fde) than if you left the original codepath.

Currently it's macOS only, but it would be easy for someone interested in supporting Windows or Linux to add an implementation to the FDE plugin.

Comment (felangel) [reply lib/src/hydrated_bloc_storage.dart:75]: @stuartmorgan yeah I talked to @AppleEducate about reverting these changes and just trying to facilitate the updates in the lower level dependencies. Is the plan to eventually merge the desktop_path_provider implementation into path_provider?

Comment (stuartmorgan-g) [reply lib/src/hydrated_bloc_storage.dart:75]: Yes, the FDE implementations of plugins from flutter/plugins will migrate into flutter/plugins as each platform stabilizes. macOS implementations will start moving soon.

Even before that though, all someone needs to do is add the FDE plugin as a dependency in their project and anything using the main plugin will Just Work, due to the way method channels are implemented. So for macOS, all someone would need to do to use your package without the changes here would be to add path_provider_fde to their project. Once Windows and Linux implementations of path_provider_fde are in place, the same would be true there.

Comment (felangel) [reply lib/src/hydrated_bloc_storage.dart:75]: Awesome, I just tested it on MacOS and created https://github.com/felangel/hydrated_bloc/pull/24. Can you please take a look?
